File: app/src/main/java/ru/alexsuvorov/paistewiki/Adapter/NewsAdapter.java
package ru.alexsuvorov.paistewiki.Adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.TextView;

import java.util.List;

import ru.alexsuvorov.paistewiki.R;
import ru.alexsuvorov.paistewiki.model.NewsMonth;
import ru.alexsuvorov.paistewiki.model.NewsPost;

public class NewsAdapter extends RecyclerView.Adapter<NewsAdapter.NewsCardViewHolder> implements View.OnClickListener {

    private List<NewsMonth> months;
    private List<NewsPost> posts;
    private AdapterView.OnItemClickListener onItemClickListener;
    private String TAG = "NewsAdapter";
    private static int posision;

    public NewsAdapter(List<NewsMonth> months) {
        this.months = months;
    }

    @Override
    public void onClick(View v) {

    }

    class NewsCardViewHolder extends RecyclerView.ViewHolder {
        CardView cv;
        TextView monthName;
        TextView postLabel;
        TextView postType;
        //TextView newsLink;

        NewsCardViewHolder(View itemView) {
            super(itemView);
            cv = itemView.findViewById(R.id.cv);
            monthName = itemView.findViewById(R.id.month_name);
            postLabel = itemView.findViewById(R.id.news_label);
            postType = itemView.findViewById(R.id.news_category);
        }
    }

    @NonNull
    @Override
    public NewsCardViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View v = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.news_item, viewGroup, false);
        return new NewsCardViewHolder(v);
    }


    public void setOnItemClickListener(AdapterView.OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    /*@Override public void onClick(final View v) {
        // Give some time to the ripple to finish the effect
        if (onItemClickListener != null) {
            new Handler().postDelayed(new Runnable() {
                @Override public void run() {
                    onItemClickListener.onItemClick(v, (ViewModel) v.getTag());
                }
            }, 200);
        }
    }*/

    /*public interface OnItemClickListener {

        void onItemClick(View view, ViewModel viewModel);

    }*/

    @Override
    public void onBindViewHolder(@NonNull NewsCardViewHolder ViewHolder, int i) {
        posision = i;
        ViewHolder.monthName.setText(months.get(i).getMonthName());
    }

    @Override
    public int getItemCount() {
        return months.size();
    }

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
    }

    public static int getMonth() {
        return posision;
    }
}

    /*
    //http://android-delight.blogspot.ru/2015/12/tablelayout-like-listview-multi-column.html
    @NonNull
    public View getView(final int position, View convertView, @NonNull ViewGroup parent) {
        View view = convertView;
        final ViewHolder holder;

        if (view == null) {
            LayoutInflater inflater = (LayoutInflater) activity
                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            assert inflater != null;
            view = inflater.inflate(row, null);
            holder = new ViewHolder();
            view.setTag(holder);
        } else {
            holder = (ViewHolder) view.getTag();
        }
        if ((news == null) || ((position + 1) > news.size()))
            return view;

        NewsMonth newsModel = news.get(position);

        holder.titleNews = view.findViewById(R.id.titleNews);
        //holder.tvMonth = view.findViewById(R.id.tvMonth);
        //holder.catNews = view.findViewById(R.id.catNews);

        /*if (holder.titleNews != null && null != newsModel.getTitle()
                && newsModel.getTitle().trim().length() > 0) {
            holder.titleNews.setText(newsModel.getTitle());
            holder.titleNews.setMovementMethod(LinkMovementMethod.getInstance());
        }*/


        /*if (holder.imgView != null) {
            if (null != objBean.getLink()
                    && objBean.getLink().trim().length() > 0) {
                Picasso.with(this.getContext().getApplicationContext())
                        .load(String.valueOf(Html.fromHtml(objBean.getLink())))
                        .resize(250, 190)
                        .placeholder(R.mipmap.ic_empty)
                        .centerCrop()
                        .into(holder.imgView);
            } else {
                holder.imgView.setImageResource(R.mipmap.ic_launcher);
            }
        }*/

        /*
        holder.btnView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Item item = getItem(position);
                Bundle bundle = new Bundle();
                assert item != null;
                bundle.putString("ID", item.getId());
                bundle.putString("Title", item.getTitle());
                bundle.putString("ImgLink", item.getLink());
                bundle.putString("Text", item.getText());
                bundle.putString("DopText", item.getDopText());

                Fragment fragment = null;
                try {
                    fragment = ArticleFragment.class.newInstance();
                } catch (InstantiationException | IllegalAccessException e) {
                    e.printStackTrace();
                }
                if (fragment != null) {
                    fragment.setArguments(bundle);
                }

                ((FragmentActivity) getContext()).getSupportFragmentManager().beginTransaction()
                        .replace(R.id.flContent, fragment)
                        .addToBackStack(null)
                        .commit();
            }
        });

        return view;

    private class ViewHolder {
        private TextView titleNews, catNews, tvMonth;
    }*/

File: app/src/main/java/ru/alexsuvorov/paistewiki/adapter/NewsAdapter.kt
package ru.alexsuvorov.paistewiki.adapter

import android.content.Context
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TableLayout
import android.widget.TableRow
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import ru.alexsuvorov.paistewiki.R
import ru.alexsuvorov.paistewiki.adapter.NewsAdapter.NewsCardViewHolder
import ru.alexsuvorov.paistewiki.db.AppDatabase
import ru.alexsuvorov.paistewiki.model.Month

class NewsAdapter(
    private val months: MutableList<Month?>,
    private val context: Context
) : RecyclerView.Adapter<NewsCardViewHolder?>() {

    private var itemClickListner: onItemClickListner? = null

    class NewsCardViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var cv: CardView? = itemView.findViewById<CardView?>(R.id.cv)
        var monthName: TextView = itemView.findViewById<TextView>(R.id.month_name)
        var tableLayout: TableLayout = itemView.findViewById<TableLayout>(R.id.postListLayout)
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): NewsCardViewHolder {
        val v = LayoutInflater.from(viewGroup.context).inflate(R.layout.item_news, viewGroup, false)
        return NewsCardViewHolder(v)
    }

    override fun onBindViewHolder(ViewHolder: NewsCardViewHolder, position: Int) {
        val db: AppDatabase = AppDatabase.Companion.getDatabase(context)
        val newsDao = db.newsDao()!!
        val monthDao = db.monthDao()!!

        val vposition = monthDao.count - position
        //Log.d(getClass().getSimpleName(),"Last position ID: "+monthDao.getLastMonthId());
        //Typeface myTypeface = Typeface.createFromAsset(context.getAssets(), "fonts/roboto_regular.ttf");
        //Log.d(getClass().getSimpleName(),"Position: "+vposition);
        val item = monthDao.getMonthById(vposition)!!

        ViewHolder.monthName.text = item.monthName
        ViewHolder.monthName.setTextColor(context.resources.getColor(R.color.black))
        val posts = newsDao.getNewsByMonthIndex(item.monthIndex.toLong())?: emptyList()
        if (posts.isNotEmpty()) {
            for (j in posts.indices) {
                val postLabel = TextView(context)
                postLabel.gravity = Gravity.START
                val layoutParams = TableRow.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.MATCH_PARENT)
                layoutParams.setMargins(0, 0, 16, 0)
                postLabel.layoutParams = layoutParams
                postLabel.textSize = 16f
                postLabel.setPadding(10, 8, 0, 8)
                postLabel.setTextColor(context.resources.getColor(R.color.black))
                //postLabel.setTypeface(myTypeface, Typeface.BOLD);
                //postLabel.setBackgroundResource(R.drawable.divider);
                postLabel.text = newsDao.getNewsByMonthIndex(item.monthIndex.toLong())!!.get(j)!!.title
                postLabel.isClickable = true

                //LEFT PICTURES
                if (posts[j]!!.category!! == "Artist News") {
                    postLabel.setCompoundDrawablesWithIntrinsicBounds(R.mipmap.ic_artist, 0, 0, 0)
                } else {
                    postLabel.setCompoundDrawablesWithIntrinsicBounds(R.mipmap.ic_cymbal_icon, 0, 0, 0)
                }

                //POSTS ON CLICK EVENS
                val data = newsDao.getNewsByMonthIndex(item.monthIndex.toLong())!!.get(j)!!.url
                postLabel.setOnClickListener(object : View.OnClickListener {
                    override fun onClick(view: View?) {
                        itemClickListner?.onClick(data)
                    }
                })

                val row = TableRow(context)
                row.addView(postLabel) // добавляем в строку столбец с именем пользователя
                row.setPadding(2, 2, 2, 2)
                ViewHolder.tableLayout.addView(row) // добавляем в таблицу новую строку
            }
        } else {
            val postLabel = TextView(context)
            postLabel.gravity = Gravity.START
            val layoutParams = TableRow.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.MATCH_PARENT)
            layoutParams.setMargins(16, 0, 16, 0)
            postLabel.layoutParams = layoutParams
            postLabel.textSize = 16f
            // postLabel.setTypeface(myTypeface, Typeface.BOLD);
            postLabel.setPadding(10, 8, 0, 8)
            postLabel.setTextColor(context.resources.getColor(R.color.black))
            //postLabel.setTypeface(myTypeface, Typeface.BOLD);
            //postLabel.setBackgroundResource(R.drawable.divider);
            postLabel.setText(R.string.no_news_yet)
            postLabel.isClickable = true
            val row = TableRow(context)
            row.addView(postLabel)
            row.setPadding(2, 2, 2, 2)
            ViewHolder.tableLayout.addView(row)
        }
    }

    override fun getItemCount(): Int {
        return months.size
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
    }

    fun setOnItemClickListner(onItemClickListner: onItemClickListner) {
        this.itemClickListner = onItemClickListner
    }

    interface onItemClickListner {
        fun onClick(str: String?)
    }
}